### PR TITLE
Fix numerical issues for Neumann inverse for KDA/GDN-2

### LINKED ISSUE #1051: [Bug]: FROST KDA produces NaN/Inf with correlated BF16 keys and safe_gate=True
## Problem

FROST KDA returns NaN/Inf for finite BF16 inputs with strongly correlated keys, even with `safe_gate=True, gate_lower_bound=-5`. Both `batch_invariant=False` and `True` are affected.

Environment: NVIDIA B300, PyTorch 2.9.0, CUDA 13.0, cuDNN Frontend 1.29.0, CuTe DSL 4.7.0.

The repro uses only generated data, one sequence (`cu_seqlens=int32 [0,8192]`), no initial state, and two heads of dimension 128. Key norms are about 0.9, beta is 1, and the activated log gates are approximately [-0.0341, -0.0121].

## Expected / actual

An independent FP64 token-by-token recurrence remains finite (`max(abs(output)) = 0.0097991`). cuDNN produces:

| batch_invariant | Nonfinite output elements |
| --- | ---: |
| False | 676592 |
| True | 676768 |

All seven returned input/parameter gradient tensors also contain NaN/Inf. Replacing the correlated keys with independent normalized random keys keeps outputs and gradients finite; forward relative RMS error is approximately 0.43%.

## Reproducer

Save as `repro.py` and run `python3 repro.py`. It prints results for both modes and ends with an assertion failure. No data files or additional model package are required.

```python
import os
os.environ.setdefault("CUBLAS_WORKSPACE_CONFIG", ":4096:8")

import torch
import torch.nn.functional as F
from cudnn.linear_attention import kimi_delta_attention

torch.set_num_threads(1)
torch.manual_seed(42)
torch.backends.cuda.matmul.allow_tf32 = False
T, H, D = 8192, 2, 128
shape = (T, H, D)

# CPU generation makes the data independent of the GPU's SM count.
q = F.normalize(torch.randn(shape), dim=-1)
direction = F.normalize(torch.randn(1, H, D), dim=-1)
k = 0.9 * F.normalize(direction + 0.002 * torch.randn(shape), dim=-1)
v, g = 0.1 * torch.randn(shape), 0.1 * torch.randn(shape)
cpu = {name: x.bfloat16() for name, x in zip(("q", "k", "v", "g"), (q, k, v, g))}
cpu.update(beta=torch.ones(T, H, dtype=torch.bfloat16),
           a_log=torch.zeros(H), dt_bias=torch.full((H, D), -5.5))
dy = torch.randn(shape).bfloat16()
assert all(torch.isfinite(x).all() for x in cpu.values())

# Independent FP64 token-by-token recurrence, with exactly the quantized inputs.
q, k, v, g, beta, a_log, dt_bias = [x.double() for x in cpu.values()]
alpha = (-5 * torch.sigmoid(a_log.exp()[:, None] * (g + dt_bias))).exp()
state = torch.zeros(H, D, D, dtype=torch.float64)
ref = []
for t in range(T):
    state = alpha[t, :, :, None] * state
    residual = v[t] - torch.einsum("hk,hkv->hv", k[t], state)
    state = state + beta[t, :, None, None] * k[t, :, :, None] * residual[:, None, :]
    ref.append(torch.einsum("hk,hkv->hv", q[t] / D**0.5, state))
ref = torch.stack(ref)
assert torch.isfinite(ref).all()
print("reference max_abs:", ref.abs().max().item(), flush=True)

passed = True
for bi in (False, True):
    xs = {name: x.cuda().requires_grad_() for name, x in cpu.items()}
    y = kimi_delta_attention(
        **xs, cu_seqlens=torch.tensor([0, T], device="cuda", dtype=torch.int32),
        initial_state=None, output_final_state=False,
        safe_gate=True, gate_lower_bound=-5.0,
        batch_invariant=bi, plan_name="kda_frost",
    )[0]
    grads = torch.autograd.grad(y, tuple(xs.values()), dy.cuda())
    torch.cuda.synchronize()
    nonfinite = {"output": int((~torch.isfinite(y)).sum())}
    nonfinite.update({"d_" + name: int((~torch.isfinite(grad)).sum())
                      for name, grad in zip(xs, grads)})
    print("batch_invariant:", bi, "nonfinite:", nonfinite, flush=True)
    passed = passed and not any(nonfinite.values())
assert passed, "cuDNN returns NaN/Inf for finite bounded inputs; FP64 reference stays finite"
```

Possible location: the BF16 intermediate casts in the 16×16 Neumann inverse in `frost/kernel/kda_prefill_f16.py`. Correlated keys appear to amplify cancellation error and destabilize the recurrent state; this is not an input NaN or an out-of-range safe-gate value.

### PR DESCRIPTION
**Pull request by jhjpark**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below.
- [x] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).

## Affected area

- FE OSS kernels or CuTeDSL

## Summary

Fix the Neumann inverse for KDA/GDN-2 for numerical stability. Thanks to @Observer007 for the idea from FlashInfer cutedsl backend.

## Summary by CodeRabbit

- **Performance and Reliability**
  - Updated GDN-2 and KDA linear-attention kernels to use blockwise 16×16 matrix inversion across prefill, recompute, summary, and backward paths.
  - Standardized inverse calculations while preserving strict lower-triangular masking behavior.

- **Testing**
  - Added regression coverage for correlated, near-parallel keys with high decay factors.
  - Expanded forward and backward parity checks for sequence lengths of 256 and 1024.
  - Added configurable tolerance scaling for numerical comparisons.

### CODE REVIEW

**Comment by jhjpark** [comment]
@cudnn-ci-bot run frost

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `37c806e714f6` against base `7ac5354191cb`.

The blockwise inverse looks sound. I checked the 4x4 -> 8x8 -> 16x16 construction against the MMA fragment layout, its use across all ten KDA/GDN-2 kernels, and the unchanged barrier participant counts. No actionable correctness or approval-blocking issue found. The numerical regression is independently reproduced on the old implementation and fixed at this head; the measured performance cost is small.

Validation: Built the native extension from this exact head. On SM100 (68 SMs), cuDNN 9.25.1, Torch 2.13.0+cu130 and DSL 4.8, the new correlated-key forward/backward tests passed all 8 selected cases. Restoring the changed production files to merge-base ae80d63a9 while retaining the new tests made all 4 KDA/GDN-2 cases fail numerically: output RMS ratios 617-1540 against the 0.02 threshold. The fixed run also checked the gradients. A further 16 selected tests passed across summary forward/backward, coarse-checkpoint recomputation and representative FP16 backward configurations; 16 redundant FP16 configurations were skipped by the existing test rule. Matched head/old/old/head FROST spot check, BF16 B=1, H=8, K=V=128, T=256 and 1024. CUDA Graph GPU time: KDA T=256 62.63-62.71 -> 63.49-63.52 us; KDA T=1024 88.53-88.73 -> 89.16-89.24 us. GDN-2 stayed at approximately 63.54 and 90.2 us. CPU graph.execute enqueue time remained approximately 32-35 us without a material regression. The baseline restores only the changed production files to the merge-base, keeping the surrounding head environment matched. All 16 performance cases passed the output/final-state oracle, rewrote poisoned outputs on captured-graph replay, and correctly changed the replay results after negating V. Timing excludes compilation and preparation. Verified the owner's existing CI mirror records this full GitHub head; no duplicate run requested. The observed Rubin failures are the same 12 native Stats NVRTC failures seen on develop-related runs, the API-index failure concerns the unrelated existing template_params entry, and the release-build job failed during repository fetch with HTTP 500. The relevant broader GPU jobs were still running when checked.

Limitations: Independent GPU evidence is targeted SM100 coverage, not a completed full cross-architecture suite. The owner's running CI remains pending evidence. The performance spot check covers forward execution at two small representative shapes, not backward performance or end-to-end model throughput.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.
